doc/fundamentals/spouts.md

```diff
@@ -0,0 +1,291 @@
+# Spouts
+
+## Introduction
+
+Spouts are a way to get streaming data from any source into Pachyderm.
+To create a spout, you need three things
+
+1. A source of streaming data, such as Kafka, nifi, rabbitMQ, etc.
+2. A containerized server.
+3. A spout pipeline specification file that uses the container.
+
+The containerized server will do four things:
+
+- connect to your source of streaming data
+- read the data
+- package it into files 
+  in a `tar` stream
+- write that `tar` stream
+  to the named pipe `/pfs/out`
+
+
+In this document, 
+we'll take you through writing the server 
+with an example using Go code
+and writing the spout pipeline specification.
+
+## Creating the containerized server
+
+To create the server,
+you'll need access to client libraries
+for your streaming data source,
+a library that can write the `tar` archive format
+(such as [Go's tar package](https://golang.org/pkg/archive/tar/) 
+or [Python's tarfile module](https://docs.python.org/3.7/library/tarfile.html)),
+and requirements for how you would like to batch your data.
+For the purposes of this document, 
+we'll assume each message in the stream
+is a single file.
+
+You need that `tar` library because,
+in spouts,
+`/pfs/out` is a named pipe. 
+This is different than in pipelines, 
+where `/pfs/out` is a directory.
+
+In the example below, 
+written in Go
+and taken from the [kafka example](https://github.com/pachyderm/pachyderm/tree/master/examples/kafka)
+in the Pachyderm repo, 
+we'll go through every step you need to take.
+
+### Import necessary libraries
+
+We'll import the libraries necessary for creating a `tar` archive data stream
+and connecting to our Kafka data source.
+
+```go
+package main
+
+import (
+	"archive/tar"
+	"context"
+	"os"
+	"strings"
+	"time"
+
+	kafka "github.com/segmentio/kafka-go"
+)
+
+```
+### Parameterizing connection information
+It's a good idea to get parameters
+for connecting to your data source 
+from the environment
+or command-line parameters. 
+That way you can connect to new data sources
+using the same container code
+without recompiling
+by just setting appropriate parameters
+in the pipeline specification.
+```Go
+func main() {
+	// Get the connection info from the ENV vars
+	host := os.Getenv("HOST")
+	port := os.Getenv("PORT")
+	topic := os.Getenv("TOPIC")
+```
+
+### Connect to the streaming data source
+
+We're creating an object
+that can be used to read from our data source.
+
+That `defer` statement is the Go way to guarantee
+that the open file will be closed
+after `main()` returns.
```

Comment:
maybe worth clarifying that `defer` runs after the code in the function calling it runs (in this case `main()`), but before the function returns

the next defer is wrapped inside an anonymous function to control when it gets run